File: surface/widgets/compose/components/widgets/primitives/Button.kt
// ============================================================
// Clef Surface Compose Widget — Button
//
// Generic action trigger rendered using Material 3 button
// composables. Supports filled, outline, text, and danger
// variants with disabled and loading states. A spinning
// indicator replaces the icon slot when loading.
//
// Adapts the button.widget spec: anatomy (root, label, icon,
// spinner), states (idle, hovered, focused, pressed, disabled,
// loading), and connect attributes (data-variant, data-size,
// data-state, role) to Compose rendering.
// ============================================================

package clef.surface.compose.components.widgets.primitives

import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

// --------------- Helpers ---------------

private fun paddingForSize(size: String): PaddingValues = when (size) {
    "sm" -> PaddingValues(horizontal = 8.dp, vertical = 4.dp)
    "lg" -> PaddingValues(horizontal = 24.dp, vertical = 12.dp)
    else -> PaddingValues(horizontal = 16.dp, vertical = 8.dp) // md
}

// --------------- Component ---------------

/**
 * Button composable that renders an action trigger with Material 3
 * styling, supporting multiple visual variants and a loading state.
 *
 * @param variant Visual variant: "filled", "outline", "text", or "danger".
 * @param size Size controlling padding: "sm", "md", or "lg".
 * @param disabled Whether the button is disabled.
 * @param loading Whether the button shows a loading spinner.
 * @param onClick Callback when the button is pressed.
 * @param modifier Compose modifier for the root element.
 * @param content Slot for the button label content.
 */
@Composable
fun Button(
    variant: String = "filled",
    size: String = "md",
    disabled: Boolean = false,
    loading: Boolean = false,
    onClick: (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    content: @Composable RowScope.() -> Unit,
) {
    val isEnabled = !disabled && !loading
    val contentPadding = paddingForSize(size)

    val dangerColors = ButtonDefaults.buttonColors(
        containerColor = MaterialTheme.colorScheme.error,
        contentColor = MaterialTheme.colorScheme.onError,
        disabledContainerColor = MaterialTheme.colorScheme.error.copy(alpha = 0.38f),
        disabledContentColor = MaterialTheme.colorScheme.onError.copy(alpha = 0.38f),
    )

    when (variant) {
        "outline" -> OutlinedButton(
            onClick = { onClick?.invoke() },
            enabled = isEnabled,
            contentPadding = contentPadding,
            modifier = modifier,
        ) {
            LoadingWrapper(loading = loading, content = content)
        }

        "text" -> TextButton(
            onClick = { onClick?.invoke() },
            enabled = isEnabled,
            contentPadding = contentPadding,
            modifier = modifier,
        ) {
            LoadingWrapper(loading = loading, content = content)
        }

        "danger" -> Button(
            onClick = { onClick?.invoke() },
            enabled = isEnabled,
            colors = dangerColors,
            contentPadding = contentPadding,
            modifier = modifier,
        ) {
            LoadingWrapper(loading = loading, content = content)
        }

        else -> Button(
            onClick = { onClick?.invoke() },
            enabled = isEnabled,
            contentPadding = contentPadding,
            modifier = modifier,
        ) {
            LoadingWrapper(loading = loading, content = content)
        }
    }
}

@Composable
private fun RowScope.LoadingWrapper(
    loading: Boolean,
    content: @Composable RowScope.() -> Unit,
) {
    if (loading) {
        CircularProgressIndicator(
            modifier = Modifier.size(16.dp),
            strokeWidth = 2.dp,
            color = LocalContentColor.current,
        )
        Spacer(modifier = Modifier.width(8.dp))
    }
    content()
}

File: surface/widgets/gtk/components/widgets/primitives/Button.ts
// ============================================================
// Clef Surface GTK Widget — Button
//
// Generic action trigger rendered using GTK4 Button with CSS
// classes for variant styling. Supports filled, outline, text,
// and danger variants with disabled and loading states.
//
// Adapts the button.widget spec: anatomy (root, label, icon,
// spinner), states (idle, hovered, focused, pressed, disabled,
// loading), and connect attributes (data-variant, data-size,
// data-state, role) to GTK4 rendering.
// ============================================================

import Gtk from 'gi://Gtk?version=4.0';

// --------------- Props ---------------

export interface ButtonProps {
  label?: string;
  variant?: 'filled' | 'outline' | 'text' | 'danger';
  size?: 'sm' | 'md' | 'lg';
  disabled?: boolean;
  loading?: boolean;
  onClick?: () => void;
}

// --------------- Component ---------------

/**
 * Creates a GTK4 Button with variant-based styling, supporting
 * filled, outline, text, and danger visual modes with a loading state.
 */
export function createButton(props: ButtonProps = {}): Gtk.Widget {
  const {
    label = '',
    variant = 'filled',
    size = 'md',
    disabled = false,
    loading = false,
    onClick,
  } = props;

  const button = new Gtk.Button();
  const isEnabled = !disabled && !loading;

  if (loading) {
    const box = new Gtk.Box({ orientation: Gtk.Orientation.HORIZONTAL, spacing: 8 });
    const spinner = new Gtk.Spinner({ spinning: true });
    box.append(spinner);
    if (label) {
      box.append(new Gtk.Label({ label }));
    }
    button.set_child(box);
  } else {
    button.set_label(label);
  }

  button.set_sensitive(isEnabled);

  // Apply variant CSS classes
  const ctx = button.get_style_context();
  switch (variant) {
    case 'filled':
      ctx.add_class('suggested-action');
      break;
    case 'outline':
      ctx.add_class('flat');
      break;
    case 'text':
      ctx.add_class('flat');
      break;
    case 'danger':
      ctx.add_class('destructive-action');
      break;
  }

  // Apply size CSS classes
  switch (size) {
    case 'sm':
      ctx.add_class('compact');
      break;
    case 'lg':
      ctx.add_class('large');
      break;
  }

  if (onClick) {
    button.connect('clicked', onClick);
  }

  return button;
}
